### PR TITLE
feat: add Event Grid telemetry dependency tracking

### LINKED ISSUE #156: Provide opt-in to enable dependency tracking
**What feature would you like to have?**
Provide opt-in to enable dependency tracking to measure performance during publishing of events.

Relates to https://github.com/tomkerkhove/k8s-event-grid-bridge/issues/182

### PR DESCRIPTION
**Pull request by stijnmoreels**
Adds our dependency telemetry tracking on the Azure Event Grid publisher instance.

Closes #156

### CODE REVIEW

**Comment by stijnmoreels** [comment]
The opt-in functionality is here implicit by passing the `ILogger` instance. It seemed unnecessary to add an additional boolean for this, no?

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs [hunk 1]
```diff
@@ -256,14 +297,26 @@ public async Task PublishManyAsync<TEvent>(IEnumerable<TEvent> events)
...
+                        if (!response.IsSuccessStatusCode)
+                        {
+                            await ThrowApplicationExceptionAsync(response);
+                        }
+                    }
+                }
+                finally
                 {
-                    await ThrowApplicationExceptionAsync(response);
+                    _logger.LogDependency("Azure Event Grid", dependencyData: subject ?? "<no-subject>", targetName: TopicEndpoint, isSuccessful: isSuccessful, measurement);
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/EventGridPublisher.cs:319]: I'm not sure if the subject or the event type should be the dependency data 🤔

**Comment by tomkerkhove** [comment]
> The opt-in functionality is here implicit by passing the `ILogger` instance. It seemed unnecessary to add an additional boolean for this, no?
> Otherwise, the dependency telemetry can be opt-out with our telemetry type filter.

I can see what you mean, but maybe it's best to make it explicit, no? End-users don't know what the ILogger is for?

**Comment by stijnmoreels** [comment]
> > The opt-in functionality is here implicit by passing the `ILogger` instance. It seemed unnecessary to add an additional boolean for this, no?
> > Otherwise, the dependency telemetry can be opt-out with our telemetry type filter.
> 
> I can see what you mean, but maybe it's best to make it explicit, no? End-users don't know how the ILogger is being used?

Yes, I see your point. Thanks for reminding me! I've changed that so it takes in a set of options.

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: I'm no expert, but is the dependency data something more info about 'identifying' the event? We can always add the event type as a property in the context, right?

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs [hunk 2]
```diff
@@ -56,22 +59,77 @@ private EventGridPublisherBuilder(Uri topicEndpoint)
...
+        /// <param name="logger">The logger instance to write dependency telemetry during the interaction with the Azure Event Grid resource.</param>
+        /// <param name="configureOptions">The additional function to configure optional settings on the <see cref="IEventGridPublisher"/>.</param>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="topicEndpoint"/> is blank.</exception>
+        /// <exception cref="UriFormatException">Thrown when the <paramref name="topicEndpoint"/> is not a correct URL.</exception>
+        /// <example>
+        /// Shows how a <see cref="EventGridPublisher"/> should be created.
+        /// <code>
+        /// EventGridPublisher myPublisher =
+        ///     EventGridPublisherBuilder
+        ///         .ForTopic("http://myTopic", logger)
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:108]: Might update this one?

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs [hunk 3]
```diff
@@ -56,22 +59,77 @@ private EventGridPublisherBuilder(Uri topicEndpoint)
...
+        ///         .UsingAuthenticationKey("myAuthenticationKey")
+        ///         .Build();
+        /// </code>
+        /// </example>
+        public static EventGridPublisherBuilder ForTopic(string topicEndpoint, ILogger logger, Action<EventGridPublisherOptions> configureOptions)
+        {
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:114]: Shouldn't we guard for a null action or null logger?

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs [hunk 4]
```diff
@@ -56,22 +59,77 @@ private EventGridPublisherBuilder(Uri topicEndpoint)
...
+        /// <code>
+        /// EventGridPublisher myPublisher =
+        ///     EventGridPublisherBuilder
+        ///         .ForTopic("http://myTopic", logger)
+        ///         .UsingAuthenticationKey("myAuthenticationKey")
+        ///         .Build();
+        /// </code>
+        /// </example>
+        public static EventGridPublisherBuilder ForTopic(string topicEndpoint, ILogger logger)
+        {
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:87]: Shouldn't we guard for a null logger?

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs [hunk 5]
```diff
@@ -84,31 +142,79 @@ public static EventGridPublisherBuilder ForTopic(string topicEndpoint)
...
+        /// <code>
+        /// EventGridPublisher myPublisher =
+        ///     EventGridPublisherBuilder
+        ///         .ForTopic("http://myTopic", logger)
+        ///         .UsingAuthenticationKey("myAuthenticationKey")
+        ///         .Build();
+        /// </code>
+        /// </example>
+        public static EventGridPublisherBuilder ForTopic(Uri topicEndpoint, ILogger logger)
+        {
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:171]: Shouldn't we guard for a null logger?

File: src/Arcus.EventGrid.Publishing/Interfaces/IEventGridPublisher.cs [hunk 6]
```diff
@@ -1,44 +1,64 @@
...
-    ///     Contract for Event Grid publisher implementations which are the result of the <see cref="IBuilder.Build" />.
+    /// Represents the contract for Azure Event Grid publisher implementations which are the result of the <see cref="EventGridPublisherBuilder" />.
     /// </summary>
     public interface IEventGridPublisher
     {
         /// <summary>
-        ///     Url of the custom Event Grid topic
+        ///  Gets the URL of the custom Azure Event Grid topic.
         /// </summary>
         string TopicEndpoint { get; }
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/Interfaces/IEventGridPublisher.cs:19]:
```suggestion

```

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs [hunk 7]
```diff
@@ -256,14 +304,29 @@ public async Task PublishManyAsync<TEvent>(IEnumerable<TEvent> events)
...
+                            await ThrowApplicationExceptionAsync(response);
+                        }
+                    }
+                }
+                finally
                 {
-                    await ThrowApplicationExceptionAsync(response);
+                    if (_options.EnableDependencyTracking)
+                    {
+                        _logger.LogDependency("Azure Event Grid", dependencyData: logSubject ?? "<no-subject>", targetName: TopicEndpoint, isSuccessful: isSuccessful, measurement);
```
Comment (tomkerkhove) [inline src/Arcus.EventGrid.Publishing/EventGridPublisher.cs:328]: I would use the event type here for now, which is more valuable

**Review by tomkerkhove — APPROVED** [review]

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: It helps better understand what was being used for that dependency, more conceptual info

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs (hunk 2 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:108]: Ah, yes, thanks!

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs (hunk 3 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:114]: Well, the logging is optional and so is the configurable options function, so I would say no.

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs (hunk 4 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:87]: Yes, thanks!

File: src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs (hunk 5 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Publishing/EventGridPublisherBuilder.cs:171]: Yes, thanks!

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs:328]: Ok, let's do that! Resolving.

**Pull request merged**